src/transformStateWithClones.js

```diff
@@ -6,8 +6,27 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      newState = { ...newState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      newState = { ...newState };
+
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    } else if (action.type === 'clear') {
+      newState = {};
+    }
+    states.push({ ...newState });
```

Comment:
[DONT REPEAT YOURSELF] - Instead of creating a new copy of the newState object in each case, you can push a copy of the object to the states array at the end of the loop cycle, but outside of the `switch` block, after the current action is processed.